pkg/core/codeset_client.go

```diff
@@ -0,0 +1,199 @@
+	}
+	cs.giteaResolver = giteaResolver
+	cs.giteaClient = giteaClient
+	return &cs, nil
+}
+
+// CreateOrg creates an Org in gitea
+func (cc *CodesetClient) CreateOrg(org string) error {
+
+	_, resp, err := cc.giteaClient.GetOrg(org)
+	if resp == nil && err != nil {
+		return errors.Wrap(err, "Failed to make get org request")
+	}
+
+	if resp.StatusCode == 200 {
+		log.Printf("Organization already exists.")
+		return nil
+	}
+
+	_, _, err = cc.giteaClient.CreateOrg(gitea.CreateOrgOption{
+		Name: org,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "failed to create org")
+	}
+	return nil
+}
+
+// create git repository with given name under given org
+func (cc *CodesetClient) CreateRepo(org, name string) error {
+	_, resp, err := cc.giteaClient.GetRepo(org, name)
+	if resp == nil && err != nil {
+		return errors.Wrap(err, "Failed to make get repo request")
+	}
+
+	if resp.StatusCode == 200 {
+		log.Printf("Application '%s' already exists under '%s'", name, org)
+		return nil
+	}
+
+	log.Printf("Creating repo '%s' under '%s'...", name, org)
+	_, _, err = cc.giteaClient.CreateOrgRepo(org, gitea.CreateRepoOption{
+		Name:          name,
+		AutoInit:      true,
+		Private:       true,
+		DefaultBranch: "main",
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to create application")
+	}
+
+	return nil
+}
+
+// Create webhook for given repository and wire it to tekton listener
+func (cc *CodesetClient) CreateRepoWebhook(org, name string) error {
+	hooks, _, err := cc.giteaClient.ListRepoHooks(org, name, gitea.ListHooksOptions{})
+	if err != nil {
+		return errors.Wrap(err, "Failed to list webhooks")
+	}
+
+	for _, hook := range hooks {
+		url := hook.Config["url"]
+		if url == StagingEventListenerURL {
+			log.Printf("Webhook for '%s' already exists", name)
+			return nil
+		}
+	}
+
+	log.Printf("Creating Webhook for '%s' under '%s'...", name, org)
+	cc.giteaClient.CreateRepoHook(org, name, gitea.CreateHookOption{
+		Active:       true,
+		BranchFilter: "*",
+		Config: map[string]string{
+			"secret":       HookSecret,
+			"http_method":  "POST",
+			"url":          StagingEventListenerURL,
+			"content_type": "json",
+		},
+		Type: "gitea",
+	})
+
+	return nil
+}
+
+// Push the code from local dir to remote repo
+func (cc *CodesetClient) GitPush(org, name, location string) error {
+	log.Printf("Pushing the code to the git repository...")
+
+	giteaURL, err := cc.giteaResolver.GetGiteaURL()
+	if err != nil {
+		return errors.Wrap(err, "Failed to resolve gitea host")
+	}
+
+	u, err := url.Parse(giteaURL)
+	if err != nil {
+		return errors.Wrap(err, "Failed to parse gitea url")
+	}
+
+	username, password, err := cc.giteaResolver.GetGiteaCredentials()
+	if err != nil {
+		return errors.Wrap(err, "Failed to resolve gitea credentials")
+	}
+
+	u.User = url.UserPassword(username, password)
+	u.Path = path.Join(u.Path, org, name)
+
+	cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf(`
```

Comment:
I do not like this, it's directly copied from original carrier based code. I'd like to use gitea api for that

Reply:
Or rather git API, not gitea, as discussed elsewhere...